nft_ingester/src/bin/slot_persister/main.rs

```diff
@@ -497,3 +559,261 @@ async fn process_slots(
         }
     }
 }
+
+async fn get_missed_slots(
+    backfill_source: Arc<BackfillSource>,
+    target_db: Arc<SlotStorage>,
+    args: Args,
+    metrics: Arc<BackfillerMetricsConfig>,
+    cancellation_token: CancellationToken,
+) {
+    /// Get missed slots in range of last_checkpoint..(last_checkpoint + SLOT_BACKFILL_OFFSET)
+    /// Returns (last_checkpoint, next_checkpoint, missed_slots)
+    ///
+    /// Below is an example of boundary calculations.
+    ///
+    /// # Examples
+    /// ```
+    /// const MISSED_SLOT_COLLECTION_OFFSET: u64 = 3000;
+    /// const MAX_SLOTS_TO_ADVANCE: u64 = 50_000;
+    /// let lower_slot_boundary: u64 = 10_000;
+    /// let higher_slot_boundary = 110_000u64 - MISSED_SLOT_COLLECTION_OFFSET;
+    /// let slot_count =
+    /// higher_slot_boundary.saturating_sub(lower_slot_boundary).min(MAX_SLOTS_TO_ADVANCE);
+    ///
+    /// assert_eq!(slot_count, 50_000u64);
+    ///
+    /// let lower_slot_boundary: u64 = 10_000;
+    /// let higher_slot_boundary = 20_000u64 - MISSED_SLOT_COLLECTION_OFFSET;
+    /// let slot_count =
+    /// higher_slot_boundary.saturating_sub(lower_slot_boundary).min(MAX_SLOTS_TO_ADVANCE);
+    ///
+    /// assert_eq!(slot_count, 7_000u64);
+    /// ```
+    async fn get_missed_slots(
+        in_mem_dumper: Arc<InMemorySlotsDumper>,
+        slots_collector: SlotsCollector<InMemorySlotsDumper, BackfillSource>,
+        target_db: Arc<SlotStorage>,
+        slot_backfill_offset: u64,
+        cancellation_token: CancellationToken,
+    ) -> Option<(u64, u64, Vec<u64>)> {
+        let lower_slot_boundary = get_checkpoint(&target_db);
+        let higher_slot_boundary = get_last_persisted_slot(target_db.clone())
+            .saturating_sub(MISSED_SLOT_COLLECTION_OFFSET);
+        let slot_backfill_offset =
+            higher_slot_boundary.saturating_sub(lower_slot_boundary).min(slot_backfill_offset);
+        let mut last_valid_key = lower_slot_boundary;
+        info!(%lower_slot_boundary, "Acquiring missed slots starting at last checkpoint");
+
+        // Store missing slots
+        let missing_slots = Arc::new(Mutex::new(Vec::new()));
+
+        info!(
+            "Starting to collect slots from {} to {}",
+            lower_slot_boundary,
+            lower_slot_boundary + slot_backfill_offset
+        );
+
+        // Collect slots from last persisted slot down to 0
+        in_mem_dumper.clear().await;
+
+        // Start slot collection
+        let _ = slots_collector
+            .collect_slots(
+                &BUBBLEGUM_PROGRAM_ID,
+                lower_slot_boundary + slot_backfill_offset,
```

Comment:
A unlikely but possible situation on a network that has really dispersed transactions and may have 0 slots with bubblegum transactions in a continuous range over the slot_backfill_offset. In this case the method will not be able to get any slots and will potentially hang infinitely. This might apply even to the solana mainnet if applied to some early days of bubblegum.

Reply:
in my experience, it didn't hang. if we didn't collect any slots in this range we just move the checkpoint forward.

Reply:
My concern is that if there are no slots we keep last_valid_key = lower_slot_boundary; and will start over the same range the next iteration